### DISCUSSION TITLE
ANE  includes a SWF version 44 which is greater than the version of the root SWF of this application: 43

### DISCUSSION THREAD

**Discussion by phitrti** [airsdk/Adobe-Runtime-Support#2480] [Q&A]
I'm dusting off an old app. It built and ran fine last week on 33.1.1.929. That's an old AIR so i started upgrading yesterday, saw that 50.2 was coming out and waited to finish. This morning, I downloaded AIR 50.2, had the latest Distriqt ANEs downloaded yesterday. I updated the app.xml namespace:
<application xmlns="http://ns.adobe.com/air/application/50.2">

And I get this error when I try to package using ADT for Android. Haven't tried iOS yet.

Included native extension 'androidx.core' includes a SWF version 44 which is greater than the version of the root SWF of this application: 43

I'm sure there is something else I need to change on my end and I'm currently digging through old information, haven't found anything yet, and I'm at a loss as to what I need to change. Note that I use a number of Distriqt ANEs and all throw the same error. 

Thoughts what I'm missing?

**Comment by ajwfrost** [comment]
This is just your SWF file needs to be at the same or greater API level than the extensions that you're loading. AIR 33.1 was associated with a SWF version of 44; if you're building with AIR 50.x then your SWF file by default should be version 50. I suspect there's a configuration setting that you've got somewhere that is changing the "swf-version" parameter that you're using in the compiler.

So -> either find and remove that (or update it to 44 or 50), or you can always use a hex editor to change the fourth byte of the file...

Note that the AIR namespace version in the app descriptor file would also need to be updated to match/exceed the SWF version, so the namespace should end with "33.1" if you're using SWF version 44, or "50.0" if you're using SWF version 50. If you're using any of the more recent app descriptor tags then of course you need to match the version that they were first released in, so you could be using "50.2" in there....

In terms of how it works:
* The AIR application descriptor version is checked when packaging up an AIR application to ensure that the XML is formed correctly, and so if you want to use tags that are introduced at a certain version, then the descriptor namespace version must match that.
* The SWF file cannot have a version above the latest one supported by an AIR namespace version, so e.g. AIR 32.0 can't be used with SWF version above 43, AIR 33.1 limits the SWF version to 44, AIR 50.x limits the SWF version to 50.
* AIR Native Extensions follow a similar rule, plus they cannot exceed the version numbers of the application that is loading them. So if you want to use an ANE that was published for AIR 33.1 with SWF version 44, you have to match/exceed this in your application.
* When the application loads, the lower of the values is used to set the compatibility in terms of what APIs and functionality is available to you. So if you had AIR namespace of 50 but the SWF version was only 44, you wouldn't be able to call any of the AS3 APIs that were introduced in AIR 50.0

Hope that helps...

**Comment by reckhoff** [comment]
Thanks Andrew for the quick response. The app descriptor is set to

<application xmlns=http://ns.adobe.com/air/application/50.2>

I’m going through all the build settings and catching up on the adt changes to see if there is something there. Problem some old setting, long deprecated that is changing it. Just need to find it…..


-Randy

From: Andrew Frost ***@***.***>
Date: Wednesday, February 22, 2023 at 8:39 AM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] ANE includes a SWF version 44 which is greater than the version of the root SWF of this application: 43 (Discussion #2480)
EXTERNAL:  This email originated from outside of the organization.  Do not click links or open attachments unless you recognize the sender and know the content is safe.


This is just your SWF file needs to be at the same or greater API level than the extensions that you're loading. AIR 33.1 was associated with a SWF version of 44; if you're building with AIR 50.x then your SWF file by default should be version 50. I suspect there's a configuration setting that you've got somewhere that is changing the "swf-version" parameter that you're using in the compiler.

So -> either find and remove that (or update it to 44 or 50), or you can always use a hex editor to change the fourth byte of the file...

Note that the AIR namespace version in the app descriptor file would also need to be updated to match/exceed the SWF version, so the namespace should end with "33.1" if you're using SWF version 44, or "50.0" if you're using SWF version 50. If you're using any of the more recent app descriptor tags then of course you need to match the version that they were first released in, so you could be using "50.2" in there....

In terms of how it works:

  *   The AIR application descriptor version is checked when packaging up an AIR application to ensure that the XML is formed correctly, and so if you want to use tags that are introduced at a certain version, then the descriptor namespace version must match that.
  *   The SWF file cannot have a version above the latest one supported by an AIR namespace version, so e.g. AIR 32.0 can't be used with SWF version above 43, AIR 33.1 limits the SWF version to 44, AIR 50.x limits the SWF version to 50.
  *   AIR Native Extensions follow a similar rule, plus they cannot exceed the version numbers of the application that is loading them. So if you want to use an ANE that was published for AIR 33.1 with SWF version 44, you have to match/exceed this in your application.
  *   When the application loads, the lower of the values is used to set the compatibility in terms of what APIs and functionality is available to you. So if you had AIR namespace of 50 but the SWF version was only 44, you wouldn't be able to call any of the AS3 APIs that were introduced in AIR 50.0

Hope that helps...

**Reply by ajwfrost** [reply]
That looks good, but it means you must be setting a "swf-version" value somewhere. Maybe in the config file (airmobile-config.xml) or whatever it is that's loading in when you compile the SWF? Or it may be an IDE setting perhaps, so the location of that would depend on what you're using for compilation... (hence the option to hand-edit the output to change the fourth byte to 0x32 i.e. v50!)

**Comment by phitrti** [comment]
I’m probably the only person on the planet still using flex. Airmobile-config.xml was it. I change the swf-version to 50 and it packaged fine.

Question- should the flex-config “target-player” value be set to something other than the “32.0” it is now?

Suggestion- in the release notes for AIR, include a section for FLEX users the changes they need to make. Not details on how to copy AIR on top of FLEX but rather the additional changes needed after that copy like swf-version and what the value should be. You could even title it “For the Lone user who still uses Flex….”

Thanks Andrew for the assist!

-Randy

From: Andrew Frost ***@***.***>
Date: Wednesday, February 22, 2023 at 8:50 AM
To: airsdk/Adobe-Runtime-Support ***@***.***>
Cc: [RCD] PHIT ***@***.***>, Author ***@***.***>
Subject: Re: [airsdk/Adobe-Runtime-Support] ANE includes a SWF version 44 which is greater than the version of the root SWF of this application: 43 (Discussion #2480)
EXTERNAL:  This email originated from outside of the organization.  Do not click links or open attachments unless you recognize the sender and know the content is safe.


That looks good, but it means you must be setting a "swf-version" value somewhere. Maybe in the config file (airmobile-config.xml) or whatever it is that's loading in when you compile the SWF? Or it may be an IDE setting perhaps, so the location of that would depend on what you're using for compilation... (hence the option to hand-edit the output to change the fourth byte to 0x32 i.e. v50!)

**Reply by ajwfrost** [reply]
> Question- should the flex-config “target-player” value be set to something other than the “32.0” it is now?

Good question! I don't think the AIR compiler uses this value, I think it's more for Flash Player targets (i.e. it looks under `frameworks/libs/player/${target-player}/playerglobal.swc` for the Flash Player AS3 APIs). For AIR, it should use `airglobal.swc` and this isn't provided under versioned folders...

> Suggestion- in the release notes for AIR, include a section for FLEX users the changes they need to make

Good idea -> have just updated this in the master document so it will be there for the next release... There are actually a lot of people still using Flex (a surprising number, in my opinion!), merging AIR SDKs into Flex is still a task that we hope to make easier...

thanks
